File: packages/sources/sdk-py/tests/golden/harness/path_utils.py
# MPLP v1.0.0 FROZEN
# Governance: MPGC

# Copyright 2025 邦士（北京）网络科技有限公司.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Path utilities for Golden Harness - Protocol-level path resolution with wildcard support.

This module provides path parsing and value extraction capabilities that are semantically
aligned with the TypeScript implementation (tests/golden/harness/ts/path-utils.ts).

Supports:
- Property access: foo.bar
- Array indexing: items[0].id
- Wildcard iteration: items[*].id (expands to items[0].id, items[1].id, ...)
- Nested wildcards: root_span.children[*].tags[*]
"""

from typing import List, Any, Union, NamedTuple
from dataclasses import dataclass


@dataclass
class PropToken:
    """Property access token (e.g., 'foo' in 'foo.bar')."""
    name: str
    type: str = "prop"


@dataclass
class IndexToken:
    """Array index token (e.g., '[0]' in 'items[0]')."""
    index: int
    type: str = "index"


@dataclass
class WildcardToken:
    """Array wildcard token (e.g., '[*]' in 'items[*]')."""
    type: str = "wildcard"


PathToken = Union[PropToken, IndexToken, WildcardToken]


class Node(NamedTuple):
    """A node in the path traversal result."""
    path: str
    value: Any


def parse_path(path: str) -> List[PathToken]:
    """
    Parse a path string into a sequence of tokens.
    
    Supports:
    - Dot notation: foo.bar.baz
    - Array indexing: items[0]
    - Wildcards: items[*]
    - Mixed: root_span.children[*].tags[0]
    
    Args:
        path: Path string to parse
        
    Returns:
        List of path tokens
        
    Raises:
        ValueError: If path syntax is invalid
        
    Examples:
        >>> parse_path("steps[*].description")
        [PropToken(name='steps'), WildcardToken(), PropToken(name='description')]
        
        >>> parse_path("root_span.children[1].name")
        [PropToken(name='root_span'), PropToken(name='children'), 
         IndexToken(index=1), PropToken(name='name')]
    """
    tokens: List[PathToken] = []
    i = 0
    n = len(path)
    current = ""
    
    def push_prop():
        nonlocal current
        if current:
            tokens.append(PropToken(name=current))
            current = ""
    
    while i < n:
        ch = path[i]
        
        if ch == ".":
            # Segment separator
            push_prop()
            i += 1
            continue
        
        if ch == "[":
            # Flush property before bracket
            push_prop()
            end = path.find("]", i)
            if end == -1:
                raise ValueError(f"Invalid path, missing ] in: {path}")
            inside = path[i + 1:end].strip()
            if inside == "*":
                tokens.append(WildcardToken())
            else:
                try:
                    idx = int(inside)
                    tokens.append(IndexToken(index=idx))
                except ValueError:
                    raise ValueError(f"Invalid array index in path: {path}")
            i = end + 1
            continue
        
        # Normal char for property name
        current += ch
        i += 1
    
    push_prop()
    return tokens


def get_value_nodes_by_path(root: Any, path: str) -> List[Node]:
    """
    Traverse a path and return all matching nodes (with actual paths).
    
    Supports wildcard expansion: steps[*].description returns all step descriptions
    with their precise paths (steps[0].description, steps[1].description, ...).
    
    Args:
        root: Root object to traverse
        path: Path string (supports wildcards)
        
    Returns:
        List of Nodes with (path, value) pairs for all matches
        
    Examples:
        >>> obj = {"steps": [{"description": "A"}, {"description": "B"}]}
        >>> nodes = get_value_nodes_by_path(obj, "steps[*].description")
        >>> [(n.path, n.value) for n in nodes]
        [('steps[0].description', 'A'), ('steps[1].description', 'B')]
    """
    tokens = parse_path(path)
    # Each step maintains a list of candidate nodes
    nodes: List[Node] = [Node(path="", value=root)]
    
    for token in tokens:
        next_nodes: List[Node] = []
        
        for node in nodes:
            value = node.value
            
            if value is None:
                continue
            
            if isinstance(token, PropToken):
                if isinstance(value, dict) and token.name in value:
                    new_path = f"{node.path}.{token.name}" if node.path else token.name
                    next_nodes.append(Node(path=new_path, value=value[token.name]))
                    
            elif isinstance(token, IndexToken):
                if isinstance(value, list) and len(value) > token.index:
                    new_path = f"{node.path}[{token.index}]"
                    next_nodes.append(Node(path=new_path, value=value[token.index]))
                    
            elif isinstance(token, WildcardToken):
                if isinstance(value, list):
                    for i, item in enumerate(value):
                        new_path = f"{node.path}[{i}]"
                        next_nodes.append(Node(path=new_path, value=item))
        
        nodes = next_nodes
    
    return nodes

File: tests/golden/harness/ts/path-utils.ts
/**
 * MPLP Protocol v1.0.0 — Frozen Specification
 * Freeze Date: 2025-12-03
 * Status: FROZEN (no breaking changes permitted)
 * Governance: MPLP Protocol Governance Committee (MPGC)
 *
 * © 2025 邦士（北京）网络科技有限公司. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * http://www.apache.org/licenses/LICENSE-2.0
 */

// tests/golden/harness/ts/path-utils.ts

export type PathToken = { type: "prop"; key: string }
    | { type: "index"; index: number }
    | { type: "wildcard" };

function parsePath(path: string): PathToken[] {
    const tokens: PathToken[] = [];
    let i = 0;
    const n = path.length;
    let current = "";

    const pushProp = () => {
        if (current.length > 0) {
            tokens.push({ type: "prop", key: current });
            current = "";
        }
    };

    while (i < n) {
        const ch = path[i];

        if (ch === ".") {
            // segment separator
            pushProp();
            i++;
            continue;
        }

        if (ch === "[") {
            // flush property before bracket
            pushProp();
            const end = path.indexOf("]", i);
            if (end === -1) {
                throw new Error(`Invalid path, missing ] in: ${path}`);
            }
            const inside = path.slice(i + 1, end).trim();
            if (inside === "*") {
                tokens.push({ type: "wildcard" });
            } else {
                const idx = Number(inside);
                if (!Number.isInteger(idx)) {
                    throw new Error(`Invalid array index in path: ${path}`);
                }
                tokens.push({ type: "index", index: idx });
            }
            i = end + 1;
            continue;
        }

        // normal char for property name
        current += ch;
        i++;
    }

    pushProp();
    return tokens;
}

/**
 * 遍历路径，返回所有匹配到的节点（带实际 path）。
 * 支持:
 *  - foo.bar
 *  - items[0].id
 *  - items[*].id
 */
export function getValueNodesByPath(
    root: any,
    path: string
): Array<{ path: string; value: any }> {
    const tokens = parsePath(path);
    // 每一步都是一组候选节点
    // 节点 = { path, value }
    let nodes: Array<{ path: string; value: any }> = [{ path: "", value: root }];

    for (const token of tokens) {
        const next: Array<{ path: string; value: any }> = [];

        for (const node of nodes) {
            const value = node.value;

            if (value === null || value === undefined) {
                continue;
            }

            if (token.type === "prop") {
                if (typeof value === "object" && !Array.isArray(value) && token.key in value) {
                    const newPath = node.path
                        ? `${node.path}.${token.key}`
                        : token.key;
                    next.push({ path: newPath, value: (value as any)[token.key] });
                }
            } else if (token.type === "index") {
                if (Array.isArray(value) && value.length > token.index) {
                    const newPath = `${node.path}[${token.index}]`;
                    next.push({ path: newPath, value: value[token.index] });
                }
            } else if (token.type === "wildcard") {
                if (Array.isArray(value)) {
                    for (let i = 0; i < value.length; i++) {
                        const newPath = `${node.path}[${i}]`;
                        next.push({ path: newPath, value: value[i] });
                    }
                }
            }
        }

        nodes = next;
    }

    return nodes;
}
